examples/consumer_rebalance_example/consumer_rebalance_example.go

```diff
@@ -0,0 +1,206 @@
+
+	run := true
+	for run == true {
+		select {
+		case sig := <-sigchan:
+			fmt.Printf("%% Caught signal %v: terminating\n", sig)
+			run = false
+		default:
+			ev := c.Poll(100)
+			if ev == nil {
+				continue
+			}
+
+			if err = processEvent(c, ev); err != nil {
+				fmt.Fprintf(os.Stderr, "Failed to process event: %s\n", err)
+			}
+
+		}
+	}
+
+	fmt.Printf("%% Closing consumer\n")
+	c.Close()
+}
+
+// processEvent processes the message/error received from the kafka Consumer's
+// Poll() method.
+func processEvent(c *kafka.Consumer, ev kafka.Event) error {
+	switch e := ev.(type) {
+
+	case *kafka.Message:
+		fmt.Printf("%% Message on %s:\n%s\n", e.TopicPartition, string(e.Value))
+
+		// Handle manual commit since enable.auto.commit is unset.
+		if err := maybeCommit(c, e.TopicPartition); err != nil {
+			return err
+		}
+
+	case kafka.Error:
+		// Errors should generally be considered informational, the client
+		// will try to automatically recover.
+		fmt.Fprintf(os.Stderr, "%% Error: %v: %v\n",
+			e.Code(), e)
+
+	default:
+		fmt.Printf("Ignored %v\n", e)
+	}
+
+	return nil
+}
+
+// maybeCommit is called for each message we receive from a Kafka topic.
+// This method can be used to apply some arbitary logic/processing to the
+// offsets, write the offsets into some external storage, and finally, to
+// decide when we want to commit already-stored offsets into Kafka.
+func maybeCommit(c *kafka.Consumer, topicPartition kafka.TopicPartition) error {
+	// Commit the already-stored offsets to Kafka whenever the offset is divisible
+	// by 10, otherwise return early.
+	// This logic is completely arbitrary. We can use any other internal or
+	// external variables to decide when we commit the already-stored offsets.
+	if topicPartition.Offset%10 != 0 {
+		return nil
+	}
+
+	commitedOffsets, err := c.Commit()
+
+	// ErrNoOffset occurs when there are no stored offsets to commit. This
+	// can happen if we haven't stored anything since the last commit.
+	// While this will never happen for this example since we call this method
+	// per-message, and thus, always have something to commit, the error
+	// handling is illustrative of how to handle it in cases we call Commit()
+	// in another way, for example, every N seconds.
+	if err != nil && err.(kafka.Error).Code() != kafka.ErrNoOffset {
+		return err
+	}
+
+	fmt.Printf("%% Commited offsets to Kafka: %v\n", commitedOffsets)
+	return nil
+}
+
+// rebalanceCallback is called on each group rebalance to assign additional
+// partitions, or remove existing partitions, from the consumer's current
+// assignment.
+//
+// The application may use this optional callback to inspect the assignment,
+// alter the initial start offset (the .Offset field of each assigned partition),
+// and read/write offsets to commit to an alternative store outside of Kafka.
+func rebalanceCallback(c *kafka.Consumer, event kafka.Event) error {
+	switch ev := event.(type) {
+	case kafka.AssignedPartitions:
```

Comment:
This needs to support incremental rebalancing too, see cooperative_consumer_example.go:

if the rebalance protocol is EAGER we use standard Assign() / Unassign(), otherwise Incremental*(). Also pay attention to AssignmentLost (which should cause an Unassign()).

Reply:
Done for the most part - I have separated the code for Eager and Cooperative protocols despite there being a very large amount of repetition. The reason for this is that this being an example, people are likely to want to just copy and paste a bunch of stuff. 
And since we can't mix protocols, they're likely to want either just Eager, or just Cooperative, and there's no need to check on every rebalance.

> Also pay attention to AssignmentLost (which should cause an Unassign()).
Do we need to call this manually in both cases or is just the comment okay?

Reply:
On AssignmentLost you will not be able to commit offsets.

Reply:
Understood, mentioned that in a comment while handling partitions revoked.